File: c-sources/arduinocomm/arduinocomm.cpp
//
// arduinocomm.cpp
//
//    implementation of the serial communication with the arduino. This class makes use of the 
//    QSerialPort class in the QT framework
//
//  modifications
//    08-jan-2024   JM    initial version
//    11-feb-2024   JM    initialise removed, now in constructor

#include "arduinocomm.h"

#include <QDebug.h>
#include <QByteArray>

#define BAUDRATE  QSerialPort::Baud115200

#define STX QByteArray::fromHex("2")
#define ETX QByteArray::fromHex("0")

// define some macro's for getting the high and low byte of a word

#define lowByte(w) ((uint8_t) ((w) & 0xff))
#define highByte(w) ((uint8_t) ((w) >> 8))

// constructor

arduinoComm::arduinoComm(QString port, int baudRate) {

  m_port = new QSerialPort();

  m_port->setPortName(port);
  m_port->open(QIODevice::ReadWrite);
  
  m_port->setBaudRate(baudRate);

  m_port->setDataBits(QSerialPort::Data8);
  m_port->setFlowControl(QSerialPort::NoFlowControl);
  m_port->setParity(QSerialPort::NoParity);
  m_port->setStopBits(QSerialPort::OneStop);

  m_port->flush();

  m_started = false;
}

// destructor
//
arduinoComm::~arduinoComm() {

  if (m_port->isOpen()) m_port->close();  
  if (m_port != NULL) delete m_port;

  return;
}

// startstop
//
//    this method start or stops the communication with the arduino

void arduinoComm::startstop(bool onoff) {

  short data = (onoff? 1 : 0);
  sendMsg('x',1,&data);
  m_started = onoff;

  // clear the data in the read and write buffer

  m_port->flush();
  m_port->readAll();

  return;
}

// isConnected
//
//    this method returns the information of the firmware in the arduino

QString arduinoComm::isConnected() {

  return QString("");
}

// isStarted
//
//    return a boolean indicating the communication with the arduino is isStarted

bool arduinoComm::isStarted() { 

  return m_started;
}

// sendMsg
//
//    this method sends a message to the arduino

void arduinoComm::sendMsg(char cmd, int n, short *data) {

  QByteArray bytesToWrite;
  encode(cmd, n, data, &bytesToWrite);
  m_port->write(bytesToWrite);

  return;
}

// rcvMsg
//
//    returns a true if a message is received
//        - cmd     : command of the message
//        - data    : received data   
//    both <cmd> and <data> are only valid if a message is received

bool arduinoComm::rcvMsg(char *cmd, int *n, short *data) {
  
  static QByteArray rcvBuffer;
  bool msgOK = false;

  int nbytes = m_port->bytesAvailable();
  rcvBuffer += m_port->read(nbytes);

  // while no STX is found, remove data from buffer
  // Note : there should be always one byte in the buffer

  while ((rcvBuffer.first(1) != STX)  && (rcvBuffer.size() > 1)) {
    rcvBuffer = rcvBuffer.sliced(1);
  }

  // STX found? than decode message

  if ((rcvBuffer.first(1) == STX) && (rcvBuffer.size() > 3)) {
    
    int lenMsg = 2 * int(rcvBuffer.at(2)) + 5;
    if (rcvBuffer.size() >= lenMsg) {
      msgOK = decode(rcvBuffer.sliced(0,lenMsg),cmd,n,data);
      rcvBuffer = rcvBuffer.sliced(lenMsg-1);                   
    }
  }
  return msgOK;
}

// encode
//
//      encodes the cmd and data to a QByteArray that can be written to the
//      serial port

void arduinoComm::encode(char cmd, int n, short *data, QByteArray *bytesToWrite) {

  int checksum = 0;

  // start of transmission

  bytesToWrite->append(STX);
  bytesToWrite->append(cmd);
  bytesToWrite->append(lowByte(n));

  // encode the data

  for (int i=0;i<n;i++) {
    checksum += data[i];
    bytesToWrite->append(highByte(data[i]));
    bytesToWrite->append(lowByte(data[i]));
  }

  // checksum and end of transmission

  bytesToWrite->append(lowByte(checksum));
  bytesToWrite->append(ETX);

  return;

}

// decode
//
//      decodes the stream of bytes in the receive buffer to a message. Returns a 
//      true if the message is correct decoded

bool arduinoComm::decode(const QByteArray rawData, char *cmd, int *n, short *data) {

  short checksumCalc = 0;

  // get the command and the length of the data

  *cmd = rawData.at(1);
  *n = rawData.at(2);

  // decode the data to integers

  for (int i=0;i<*n;i++) {
    
    uint8_t highByte = rawData.at(3+2*i);
    uint8_t lowByte = rawData.at(4+2*i); 
    data[i] = (highByte << 8) + lowByte;
    checksumCalc += data[i];
  }
  checksumCalc = lowByte(checksumCalc);

  // get the checksum and return a true if it is correct

  uint8_t checksumMsg = rawData.at(3 + 2 * (*n));
 
  return (checksumMsg == checksumCalc);
}

File: python/arduinocomm.py
#
# arduinocomm.py
#
#   this is the python implementation of the <arduinoComm> class, a class for communication with the
#   arduino
#
# modifications
#   07-nov-2023  JM   initial version

from PySide6.QtSerialPort import QSerialPort
from PySide6.QtCore import QByteArray, QIODevice
import struct

class arduinoComm :

  # constructor, nothing is done

  def __init__(self,port,baudrate = 230400) :
    
    self.rcvBuffer = QByteArray()
    self.m_started = False

    self.m_port = QSerialPort()

    self.m_port.setPortName(port)
    self.m_port.open(QIODevice.ReadWrite)
  
    self.m_port.setBaudRate(baudrate)
    self.m_port.setDataBits(QSerialPort.Data8)
    self.m_port.setFlowControl(QSerialPort.NoFlowControl)
    self.m_port.setParity(QSerialPort.NoParity)
    self.m_port.setStopBits(QSerialPort.OneStop)
    
    self.m_port.flush()
    
    return
  
  # destructor

  def destroy(self) :
  
    if self.m_port.isOpen() :
      self.m_port.close()
    self.m_started = False
    return
  
  # startstop
  # 
  # start or stop the module

  def startstop(self,onoff) :

    if onoff :
      self.m_started = True
      self.sendMsg("x",[0x01])
    else :
      self.m_started = False
      self.sendMsg("x",[0x00])

    # and clear everything, write, read and receive buffer

    self.m_port.flush()  
    self.m_port.readAll()

    self.rcvBuffer = QByteArray()

    return
  
  # isConneected

  def isConnected(self) :

    msgOK = False
    cmd = ""
    i = 0 
    
    # because there could be data into the buffer, we flush the port before getting
    # the info (NOTE: this doesn't work)

    self.m_port.flush()
    self.rcvBuffer = QByteArray()

    self.sendMsg('v',[])
    while ( i < 5) & (cmd != 'V') :
      msgOK,cmd,data = self.rcvMsg()
      print(cmd)

      i = i + 1

    dataStr = ''
    for f in data :
      dataStr = dataStr + chr(f)
    
    return dataStr
  
  # isStarted

  def isStarted(self) :
    
    return self.m_started
    
  # sendMsg
  #
  #   sends a command with data to the serial port after encoding it
  
  def sendMsg(self,cmd,data) :

    outputData = encode(cmd,data)
    self.m_port.write(outputData)
    
    return

  # rcvMsg
  #
  # read the message, returns TRUE is a message is received

  def rcvMsg(self) :
  
    msgOK = False
    cmd = ''
    data = []
    
    # read the incoming bytes to the buffer
    
    nbytes = self.m_port.bytesAvailable()
    self.rcvBuffer += self.m_port.read(nbytes)

    # find the STX message, remove data from buffer until found
    # Note : there should be at least one byte in the buffer
    
    if (len(self.rcvBuffer) > 0 ) : 
      
      while ((self.rcvBuffer[0] != b'\x02') and (len(self.rcvBuffer) > 1)) :
        self.rcvBuffer = self.rcvBuffer[1:]

      # STX found, then decode message
        
      if ((self.rcvBuffer[0] == b'\x02') and (len(self.rcvBuffer) > 3)) :

        lenMsg = 2 * ord(self.rcvBuffer[2].decode()) + 5
        if (len(self.rcvBuffer) >= lenMsg) :
        
          cmd,data,msgOK = decode(self.rcvBuffer[0:lenMsg])
          if msgOK : self.rcvBuffer = self.rcvBuffer[(lenMsg-1):]

    return msgOK,cmd,data


# encode
#
#   this routine encodes the command and the data to a byte array that can be send to the arduino

def encode(cmd,data) :

  STX = 0x02
  EOT = 0x00

  outputData = bytearray()
  
  outputData += STX.to_bytes(1,'big')
  outputData += bytes(cmd,'ascii')
  outputData += len(data).to_bytes(1,'big')

  for num in data :
    outputData += num.to_bytes(2,'big')

  checksum = sum(data) & 0x00FF
  outputData += checksum.to_bytes(1,'big')
  outputData += EOT.to_bytes(1,'big')

  return outputData

# decode
#
#    decodes a message

def decode(inputData) :

  # get command and lenght of the message

  cmd = inputData[1].decode()
  lenMsg = ord(inputData[2].decode())
  
  # get the data

  dataSlice = inputData[3:-2]
  data = [int.from_bytes(dataSlice[i:i+2], byteorder = "big",signed = True) for i in range(0, len(dataSlice), 2)]
  
  # checksum OK ?
  
  checksum = inputData[-2:-1]
  msgOK = (checksum == (sum(data) & 0x00FF).to_bytes(1,'big'))

  return cmd,data,msgOK